crates/commitlog/src/payload.rs

```diff
@@ -53,3 +81,383 @@ impl<const N: usize> Decoder for ArrayDecoder<N> {
         reader.get_array()
     }
 }
+
+bitflags! {
+    #[derive(Clone, Copy)]
+    pub struct Flags: u8 {
+        const HAVE_INPUTS    = 0b10000000;
+        const HAVE_OUTPUTS   = 0b01000000;
+        const HAVE_MUTATIONS = 0b00100000;
+    }
+}
+
+/// The canonical payload format of a [`crate::Commitlog`].
+///
+/// This type may eventually be defined in the core datastore crate.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Txdata {
+    pub inputs: Option<Inputs>,
+    pub outputs: Option<Outputs>,
+    pub mutations: Option<Mutations>,
+}
+
+impl Txdata {
+    pub fn encode(&self, buf: &mut impl BufWriter) {
+        let mut flags = Flags::empty();
+        flags.set(Flags::HAVE_INPUTS, self.inputs.is_some());
+        flags.set(Flags::HAVE_OUTPUTS, self.outputs.is_some());
+        flags.set(Flags::HAVE_MUTATIONS, self.mutations.is_some());
+
+        buf.put_u8(flags.bits());
+        if let Some(inputs) = &self.inputs {
+            inputs.encode(buf);
+        }
+        if let Some(outputs) = &self.outputs {
+            outputs.encode(buf);
+        }
+        if let Some(mutations) = &self.mutations {
+            mutations.encode(buf)
+        }
+    }
+
+    pub fn decode<'a, D, R>(row_decoder: &D, reader: &mut R) -> Result<Self, DecodeError>
+    where
+        D: RowDecoder,
+        R: BufReader<'a>,
+    {
+        let flags = Flags::from_bits_retain(reader.get_u8()?);
+        let inputs = flags
+            .contains(Flags::HAVE_INPUTS)
+            .then(|| Inputs::decode(reader))
+            .transpose()?;
+        let outputs = flags
+            .contains(Flags::HAVE_OUTPUTS)
+            .then(|| Outputs::decode(reader))
+            .transpose()?;
+        let mutations = flags
+            .contains(Flags::HAVE_MUTATIONS)
+            .then(|| Mutations::decode(row_decoder, reader))
+            .transpose()?;
+
+        Ok(Self {
+            inputs,
+            outputs,
+            mutations,
+        })
+    }
+}
+
+/// The inputs of a transaction, i.e. the name and arguments of a reducer call.
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct Inputs {
+    pub reducer_name: Arc<Varchar<255>>,
+    pub reducer_args: Arc<[u8]>,
+    // TODO: `reducer_args` should be a `ProductValue`, which
+    // requires `Decoder` to be able to resolve the argument
+    // type by `reducer_name`.
+    //
+    // We can do that once modules are stored in the database
+    // itself, or the database is otherwise able to load the
+    // module info.
+    //
+    // At that point, we can also remove the length prefix of
+    // `Inputs`.
+}
+
+impl Inputs {
+    pub fn encode(&self, buf: &mut impl BufWriter) {
+        let slen = self.reducer_name.len() as u8;
+        let len = /* slen */ 1 + slen as usize + self.reducer_args.len();
```

Comment:
What's the deal with the `/* slen */`? Is that related to the above comment about removing the length prefix?

Reply:
It’s the one byte length of the string, mainly `reducer_name`. Which led me to introducing `Varchar` so that encoding remains infallible.

We could instead do one of the following:

- Use a varint
- Mandate that traversals must bring a way to resolve the type of the reducer args, so that the name can come after the args and we don’t need to store the length 
- Derive a stable numerical reducer id, which may however only be stable for the lifetime of the module revision

Reply:
We also likely want to settle on some refcounted, small string representation in core, so we can avoid copying the same strings over and over again.